### DISCUSSION TITLE
Dynamic atom with async function causing infinite loop

### DISCUSSION THREAD

**Discussion by coodoo** [pmndrs/jotai#1553] [Q&A]
I've come up with this minimum test case that when creating dynamic atom with async function will cause react to infinitely rerender, I tried `useCallback` on the `fn` and `useMemo` on the `bad atom` but to no avail, wondering if this is expected or I'm doing something wrong? Thanks.

```js
const ok = atom(async get => Promise.resolve(8888))

const fn = async get => {
	return await Promise.resolve(`123`)
}

const App = () => {

	const bad = useMemo(() => atom(fn), [])

	// this caused infinite loop
	const [foo] = useAtom(bad)

	// this ran fine
	// const [foo] = useAtom(ok)

	Log.debug(`blah:`, foo)

}
```

**Comment by dai-shi** [comment]
useMemo refreshes when the component suspends. (This might be fixed in the future version of React.)
For now, what you can do is to add a Suspense boundary between a parent component with  `useMemo` and a child component with `useAtom`.

**Reply by coodoo** [reply]
@dai-shi Thanks for the quick reply, I created a contrived sample here according to your suggestion but still no luck, mind having a look?

https://codesandbox.io/s/dawn-sound-qv42yo?file=/src/App.js

**Reply by dai-shi** [reply]
Here you go: https://codesandbox.io/s/nifty-voice-g6seto?file=/src/PanelWrapper.js

**Reply by coodoo** [reply]
Ah ha! Now I see what you meant and it made perfect sense, adding an extra layer of Suspense to wrap `Panel` so that when it's `useAtom` invoke the `read` function and it's still working it will throw a `Promise` only to be caught by the nearest `Suspense`, without that the app simply goes into infinite loop!

Btw, this might be a good addition to the [`performance.md` draft](https://github.com/pmndrs/jotai/pull/1208), at least until react fixed this `useMemo` refreshing thing.

Anyway, thanks for the quick help, you rock! 🙌

**Comment by tien** [comment]
Pardon my ignorant, but will this be fixed in the upcoming React 19? Would be useful as I think this is a common use case:

```ts
const Component = (props) => {
  const value = useAtomValue(useMemo(() => Promise.all(props.someArray.map(toAtom)), []));
}
```

**Reply by dai-shi** [reply]
As far as I understand, the behavior doesn't change with 19. But you can try experimenting it. I'd recommend doing so without Jotai.

**Reply by tien** [reply]
Yeah, seem like you would still need to cache the promise somehow :v
